// @nativescript/core/ui/image/symbol-effects-common.js
export var ImageSymbolEffects;
(function (ImageSymbolEffects) {
    ImageSymbolEffects["Appear"] = "appear";
    ImageSymbolEffects["AppearUp"] = "appearUp";
    ImageSymbolEffects["AppearDown"] = "appearDown";
    ImageSymbolEffects["Bounce"] = "bounce";
    ImageSymbolEffects["BounceUp"] = "bounceUp";
    ImageSymbolEffects["BounceDown"] = "bounceDown";
    ImageSymbolEffects["Disappear"] = "disappear";
    ImageSymbolEffects["DisappearDown"] = "disappearDown";
    ImageSymbolEffects["DisappearUp"] = "disappearUp";
    ImageSymbolEffects["Pulse"] = "pulse";
    ImageSymbolEffects["Scale"] = "scale";
    ImageSymbolEffects["ScaleDown"] = "scaleDown";
    ImageSymbolEffects["ScaleUp"] = "scaleUp";
    ImageSymbolEffects["VariableColor"] = "variableColor";
    ImageSymbolEffects["Breathe"] = "breathe";
    ImageSymbolEffects["BreathePlain"] = "breathePlain";
    ImageSymbolEffects["BreathePulse"] = "breathePulse";
    ImageSymbolEffects["Rotate"] = "rotate";
    ImageSymbolEffects["RotateClockwise"] = "rotateClockwise";
    ImageSymbolEffects["RotateCounterClockwise"] = "rotateCounterClockwise";
    ImageSymbolEffects["Wiggle"] = "wiggle";
    ImageSymbolEffects["WiggleBackward"] = "wiggleBackward";
    ImageSymbolEffects["WiggleClockwise"] = "wiggleClockwise";
    ImageSymbolEffects["WiggleCounterClockwise"] = "wiggleCounterClockwise";
    ImageSymbolEffects["WiggleDown"] = "wiggleDown";
    ImageSymbolEffects["WiggleForward"] = "wiggleForward";
    ImageSymbolEffects["WiggleUp"] = "wiggleUp";
    ImageSymbolEffects["WiggleLeft"] = "wiggleLeft";
    ImageSymbolEffects["WiggleRight"] = "wiggleRight";
})(ImageSymbolEffects || (ImageSymbolEffects = {}));
export class ImageSymbolEffectCommon {
}
